driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -382,23 +404,11 @@ private void retry(final boolean retryCurrent, ConsistencyLevel newConsistencyLe
             if (newConsistencyLevel != null)
                 this.retryConsistencyLevel = newConsistencyLevel;
 
-            // We should not retry on the current thread as this will be an IO thread.
-            manager.executor().execute(new Runnable() {
-                @Override
-                public void run() {
-                    if (queryStateRef.get().isCancelled())
-                        return;
-                    try {
-                        if (retryCurrent) {
-                            if (query(h))
-                                return;
-                        }
-                        sendRequest();
-                    } catch (Exception e) {
-                        setFinalException(null, new DriverInternalError("Unexpected exception while retrying query", e));
-                    }
-                }
-            });
+            if (queryStateRef.get().isCancelled())
+                return;
+
+            if (!retryCurrent || !query(h))
```

Comment:
Yes. There's nothing blocking or compute-intensive on that path anymore: if the pool is busy or the keyspace needs to be changed, that will be handled asynchronously via the future returned from borrow. So I figured the extra indirection wasn't worth it anymore.